corehq/apps/reports/standard/cases/case_data.py

```diff
@@ -826,16 +826,29 @@ def post(self, request, *args, **kwargs):
 def soft_delete_cases_and_forms(request, domain, case_delete_list, form_delete_list, main_case_name=None):
     error = False
     msg = _("{}, its related subcases and submission forms were deleted successfully.").format(main_case_name)
-    for form in form_delete_list:
-        if archive_form(request, domain, form, is_case_delete=True):
-            form_instance = XFormInstance.objects.get_form(form, domain)
-            form_instance.soft_delete()
-        else:
+
+    def archive_and_delete_forms(form_list):
+        for form in form_list:
+            if archive_form(request, domain, form, is_case_delete=True):
+                form_instance = XFormInstance.objects.get_form(form, domain)
+                form_instance.soft_delete()
+            else:
+                raise FormArchiveError(form.form_id)
```

Comment:
Raising an error hear means that some forms in `form_list` may get deleted while others (all those not processed at the time the error is raised) may not, which would leave the system in a state where (to the user) the operation appeared to fail, but some things got deleted. This does not seem like an intuitive outcome. I think the operation should either succeed with all forms deleted or fail with no forms deleted. Is that possible?

Reply:
whoops yea that wouldn't be good. I made sure all forms are archived properly first before moving onto the soft deletion (and unarchives previously archives forms if it fails in the middle of that).

Reply:
Cool. Would it be hard to add a test for that functionality?